src/middlewared/middlewared/plugins/cluster_linux/ctdb_ips.py

```diff
@@ -0,0 +1,193 @@
+            # healthy. So we do the following:
+            #   1. if the ctdb shared volume private ip file exists
+            #       then assume that this isn't the first peer
+            #       being added to the cluster and check the ctdb
+            #       daemon for the cluster health.
+            #   2. elif the ctdb shared volume private ip doesnt
+            #       exist then assume this is the first peer being
+            #       added to the cluster and skip the cluster health
+            #       check.
+            if ctdb_file.exists() and self.middleware.call_sync('service.started', 'ctdb'):
+                if not self.middleware.call_sync('ctdb.general.healthy'):
+                    raise CallError('ctdb cluster is not healthy, not updating private ip file')
+
+        # create the ctdb shared volume file
+        # ignoring if it's already there
+        try:
+            ctdb_file.touch(exist_ok=True)
+        except Exception as e:
+            raise CallError(f'Failed creating {ctdb_file} with error: {e}')
+
+        # we need to make sure the local etc file is symlinked
+        # to the ctdb shared volume so all nodes in the cluster
+        # have the same config
+        symlink_it = delete_it = False
+        if etc_file.exists():
+            if not etc_file.is_symlink():
+                # delete it since we're symlinking it
+                delete_it = True
+            elif etc_file.resolve() != ctdb_file:
+                # means it's a symlink but not to the ctdb
+                # shared volume ip file
+                delete_it = True
+        else:
+            symlink_it = True
+
+        # delete the file
+        if delete_it:
+            try:
+                etc_file.unlink()
+            except Exception as e:
+                raise CallError(f'Failed deleting {etc_file} with error: {e}')
+
+        # symlink the file
+        if symlink_it:
+            try:
+                etc_file.symlink_to(ctdb_file)
+            except Exception as e:
+                raise CallError(f'Failed symlinking {etc_file} to {ctdb_file} with error: {e}')
+
+        if create:
+            # in the case of adding a node (private or public),
+            # it _MUST_ be added to the end of the file always
+            with open(ctdb_file, 'a') as f:
+                if is_private:
+                    f.write(data['ip'] + '\n')
+                else:
+                    entry = f'{data["ip"]}/{data["netmask"]} {data["interface"]}'
+                    f.write(entry + '\n')
+        else:
+            if is_private:
+                address = data['address']
+            else:
+                address = data["public_ip"]
+
+            find_entry = address if not enable else '#' + address
+
+            # no matter what we're doing at this point,
+            # we're going to have to read, truncate,
+            # and rewrite the contents.
+            with open(ctdb_file, 'r+') as f:
```

Comment:
Didnt we had a python issue where it couldnt open a file for read and write this way? and we had to use os.open()?

Reply:
The `write_if_changed` method has this problem but only because the `etc` plugin doesn't guarantee the file is created when you call it.

At the beginning of this method, we always `file.touch(exist_ok=True)` before we get here. So this is the way of opening the file for reading and writing and not truncating if it exists (which it should always).

Reply:
Ah, makes sense, thanks!